<a href="https://colab.research.google.com/github/enzoalvino00/Analitica-Empresarial-Integrada/blob/main/Unidad_1_Valor_del_dato/Semana_01_Empresa_analitica/Lab_D1/Analitica_Empresarial_Integrada_Lab03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


---
# 🏛️ **TECSUP 2026-II  -  ANALÍTICA EMPRESARIAL INTEGRADA**
## 🔬 **LABORATORIO DIRIGIDO N.° 03** : **La métrica que importa**


| | |
|---|---|
| **Docente** | Pilar Rocío Sayán Mejía |
| **Laboratorio** | Diagnóstico de madurez analítica de una agroexportadora peruana |
| **N.° de laboratorio** |  3 |
| **Tema** | La Empresa Analítica en la Era de la IA |
| **Plan curricular** | Big Data y Ciencia de Datos · Código C28 · 2026-I |
| **Duración** | 200 minutos (6:50 p.m. – 10:10 p.m., con receso de 15 minutos) |
| **Modalidad** | Grupal. Las respuestas del foro son individuales. |

---

**Estudiante:** _Enzo Gabriel ALvino Carhuaricra_ &nbsp;&nbsp;&nbsp; **Sección:** __6C28__ &nbsp;&nbsp;&nbsp; **Fecha:** 02/09/2026

## Agenda de laboratorio — 7:00 p. m. a 10:10 p. m.

**Duración total:** 190 minutos · **Receso:** 15 minutos · **Trabajo efectivo:** 175 minutos.

| Horario | Tiempo | Desarrollo |
|---|---:|---|
| 7:00–7:10 | 10 min | Apertura del caso y presentación del problema de medición. |
| 7:10–7:35 | 25 min | Actividad 1. Revisión de conceptos: del dato a la decisión. |
| 7:35–8:00 | 25 min | Actividad 2. Descarga desde UCI, auditoría y limpieza documentada. |
| 8:00–8:30 | 30 min | Actividad 2. Periodo comparable, recurrencia y North Star. Reto 1. |
| 8:30–8:45 | 15 min | **RECESO** |
| 8:45–9:15 | 30 min | Actividad 2. Árbol de métricas, guardrails, Polars y DuckDB. |
| 9:15–9:35 | 20 min | Actividad 2. Tablero de decisión en Plotly. Reto 2. |
| 9:35–10:00 | 25 min | **Reto de aplicación y retroalimentación.** Ejercicios 1 a 5. |
| 10:00–10:10 | 10 min | Diccionario de KPI, informe ejecutivo y ticket de salida. |

> **Regla de trabajo:** no avance de bloque sin registrar la interpretación solicitada. El objetivo no es ejecutar celdas, sino convertir datos en evidencia para una decisión.


## Actividad 1 — Revisión de conceptos: del dato a la decisión (25 minutos)

**Propósito.** Establecer con precisión el vocabulario de medición antes de programar. La confusión entre dato, métrica, indicador y KPI constituye la causa más frecuente de tableros que no sustentan ninguna decisión.

**Instrucciones.** Complete la tabla con definiciones elaboradas con sus propias palabras. No se admite la reproducción literal de fuentes externas ni de sistemas generativos. La columna de la derecha contiene una pregunta de apoyo: si su definición permite responderla, la definición es suficiente; si no lo permite, corríjala antes de continuar.

**Evidencia esperada.** Tabla completa con las definiciones registradas.

| **Concepto** | **Definición con mis palabras** | **Ejemplo del caso** |
|---|---|---|
| **Dato** | Es un valor que se obtiene directamente de la información de la empresa, pero por sí solo no nos dice si el negocio está funcionando bien o mal. | **10 000 usuarios registrados**. |
| **Métrica** | Es un dato que hemos calculado para medir alguna característica o comportamiento del negocio. | **Tasa de renovación = 240 ÷ 300 = 80 %**. |
| **Indicador** | Es una métrica a la que le damos un significado al compararla con un nivel esperado o una referencia. | La **retención de 80 %** está por debajo del **87 % esperado**. |
| **KPI** | Es un indicador que la empresa considera importante porque está relacionado directamente con un objetivo estratégico y ayuda a tomar decisiones. | **Tasa de retención mensual**, porque permite controlar si los clientes siguen permaneciendo en la empresa. |
| **Meta** | Es el resultado que queremos alcanzar en un determinado periodo. | Mantener una **tasa de retención ≥ 87 %** al cierre del semestre. |
| **North Star** | Es la métrica principal que muestra si la empresa está avanzando hacia el objetivo que realmente genera valor. | **Pedidos entregados a compradores recurrentes por mes**. |
| **Driver** | Es un factor que puede hacer que la North Star aumente o disminuya. | **Frecuencia de compra por recurrente**, porque puede influir directamente en la cantidad de pedidos recurrentes. |
| **Guardrail** | Es un límite que nos ayuda a crecer sin perjudicar otros aspectos importantes del negocio. | **Margen de contribución ≥ 20 %**, para evitar crecer sacrificando la rentabilidad. |

**Criterio de cierre.** No se avanza al desarrollo práctico mientras existan conceptos sin definición registrada.


## Actividad 2 — Desarrollo práctico y ejecución

### Presentación del caso

Una empresa minorista en línea del Reino Unido cuenta con un registro histórico de transacciones que incluye facturas, productos, cantidades, fechas, precios unitarios, clientes y países. La dirección necesita transformar estos registros en un sistema de métricas que permita distinguir crecimiento útil de crecimiento aparente. Para ello, no basta con observar ventas acumuladas o cantidad de clientes: es necesario identificar qué indicadores representan valor recurrente para el cliente y cuáles pueden orientar una decisión empresarial.

Durante el laboratorio, el equipo trabajará con el conjunto Online Retail del UCI Machine Learning Repository. A partir de los datos reales, deberá auditar y preparar las transacciones, identificar clientes recurrentes, proponer y justificar una North Star, descomponerla en drivers y guardrails, construir un diccionario de KPI y elaborar un tablero de decisión en Plotly. El análisis deberá terminar con hallazgos cuantitativos y una acción empresarial concreta, diferenciando en todo momento los valores observados en la base de las metas o umbrales académicos propuestos para el ejercicio.


## Paso 1 — Preparación reproducible del entorno


In [1]:
# Instala las versiones necesarias para garantizar un entorno reproducible
%pip install -q ucimlrepo==0.0.7 polars==1.17.1 duckdb==1.1.3

# Librerías principales para análisis, consultas y visualización
import polars as pl  ## Manipulación y transformación de datos
import duckdb  ## Consultas SQL sobre los datos
import pandas as pd  ## Recibir los datos descargados desde UCI
import plotly.express as px  ## Crear gráficos de forma sencilla
import plotly.graph_objects as go  ## Crear gráficos personalizados
from IPython.display import display  ## Mostrar tablas y resultados

# Configuración de visualización de tablas en Polars
pl.Config.set_tbl_rows(20)  ## Muestra hasta 20 filas
pl.Config.set_tbl_width_chars(160)  ## Define el ancho máximo de la tabla

# Formato de fecha utilizado por el dataset
FORMATO_FECHA = "%m/%d/%Y %H:%M"  ## Mes/día/año hora:minuto

# Verifica que las versiones instaladas sean las esperadas
print("polars:", pl.__version__, "| duckdb:", duckdb.__version__)

# Confirma que el entorno está preparado
print("Entorno listo.")


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.4/36.4 MB 42.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.2/20.2 MB 51.4 MB/s eta 0:00:00
polars: 1.17.1 | duckdb: 1.1.3
Entorno listo.


## Paso 2 — Descarga de datos reales desde UCI

El conjunto **Online Retail** contiene transacciones de una empresa minorista en línea registrada en el Reino Unido entre diciembre de 2010 y diciembre de 2011. Los códigos de factura que empiezan con `C` representan cancelaciones. No se generan registros artificiales.


In [2]:
from ucimlrepo import fetch_ucirepo

# Descarga el dataset Online Retail desde el repositorio UCI
online_retail = fetch_ucirepo(id=352)

# Convierte los datos descargados de Pandas a Polars
df = pl.from_pandas(online_retail.data.original)  ## La descarga llega en Pandas y se pasa a Polars

# Muestra el nombre del dataset
print("Dataset:", online_retail.metadata.get("name"))

# Muestra la cantidad de filas y columnas del dataset
print("Filas y columnas:", df.shape)

# Muestra los nombres de las columnas disponibles
print("Columnas:", df.columns)

# Muestra las primeras filas para revisar la estructura de los datos
display(df.head())

Dataset: Online Retail
Filas y columnas: (541909, 8)
Columnas: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
str,str,str,i64,str,f64,f64,str
"""536365""","""85123A""","""WHITE HANGING HEART T-LIGHT HO…",6,"""12/1/2010 8:26""",2.55,17850.0,"""United Kingdom"""
"""536365""","""71053""","""WHITE METAL LANTERN""",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""
"""536365""","""84406B""","""CREAM CUPID HEARTS COAT HANGER""",8,"""12/1/2010 8:26""",2.75,17850.0,"""United Kingdom"""
"""536365""","""84029G""","""KNITTED UNION FLAG HOT WATER B…",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""
"""536365""","""84029E""","""RED WOOLLY HOTTIE WHITE HEART.""",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""


### **Estructura y dimensiones del dataset Online Retail**

- **Dataset:** Online Retail
- **Filas y columnas:** 541 909 filas y 8 columnas
- **Columnas:** `InvoiceNo`, `StockCode`, `Description`, `Quantity`, `InvoiceDate`, `UnitPrice`, `CustomerID` y `Country`
- **Primeras 5 filas:** Permiten comprobar que los registros y tipos de datos se cargaron correctamente.

### Control de trazabilidad
Registre: número de filas, columnas, rango de fechas y países presentes.

**Respuesta:** Se registró el número de filas, columnas, rango de fechas y países presentes en el dataset para mantener la trazabilidad de la información utilizada en el análisis.


## **Paso 3 — Auditoría inicial y limpieza documentada**


In [3]:
# Prepara las columnas del dataset con los tipos de datos necesarios para el análisis
df = df.with_columns([
    pl.col("InvoiceNo").cast(pl.Utf8),  ## Convierte el número de factura a texto
    pl.col("CustomerID").cast(pl.Utf8),  ## Convierte el identificador del cliente a texto
    pl.col("Quantity").cast(pl.Float64, strict=False),  ## Convierte la cantidad a número decimal
    pl.col("UnitPrice").cast(pl.Float64, strict=False),  ## Convierte el precio unitario a número decimal
    pl.col("InvoiceDate").cast(pl.Utf8)  ## Convierte inicialmente la fecha a texto
      .str.to_datetime(format=FORMATO_FECHA, strict=False)  ## Convierte el texto al formato fecha-hora definido
      .alias("InvoiceDate"),  ## Mantiene el nombre original de la columna
])

# Crea variables necesarias para identificar cancelaciones y calcular el valor de cada línea
df = df.with_columns([
    pl.col("InvoiceNo").str.to_uppercase()  ## Convierte las facturas a mayúsculas
      .str.starts_with("C")  ## Comprueba si la factura comienza con "C"
      .alias("es_cancelacion"),  ## Guarda el resultado como indicador de cancelación

    (pl.col("Quantity") * pl.col("UnitPrice")).alias("importe_linea"),  ## Calcula el importe de cada línea
])

# Verifica que las fechas se hayan convertido correctamente
nulas = df["InvoiceDate"].null_count()  ## Cuenta las fechas que quedaron como valores nulos

print("Fechas que no pudieron parsearse:", nulas)  ## Muestra el resultado del control

# Detiene el proceso si alguna fecha no pudo interpretarse correctamente
if nulas:
    raise ValueError("El formato declarado en FORMATO_FECHA no corresponde a la fuente.")

# Define los valores esperados para comprobar que la fuente original no haya cambiado
FILAS_ESPERADAS = 541_909  ## Cantidad de filas esperada
FACTURAS_ESPERADAS = 25_900  ## Cantidad de facturas distintas esperada

# Verifica que el dataset tenga la cantidad de filas esperada
if df.height != FILAS_ESPERADAS:  ## Compara las filas actuales con las filas esperadas
    raise ValueError(  ## Detiene el código si las cantidades no coinciden
        f"La fuente cambio: se esperaban {FILAS_ESPERADAS:,} filas y llegaron {df.height:,}. "  ## Muestra los valores esperados y encontrados
        "Avise al docente antes de continuar."  ## Indica qué hacer si la fuente cambió
    )

# Verifica que el dataset tenga la cantidad de facturas esperada
if df["InvoiceNo"].n_unique() != FACTURAS_ESPERADAS:  ## Cuenta las facturas diferentes y las compara con las esperadas
    raise ValueError(  ## Detiene el código si las cantidades no coinciden
        f"La fuente cambio: se esperaban {FACTURAS_ESPERADAS:,} facturas distintas "  ## Muestra la cantidad de facturas esperada
        f"y llegaron {df['InvoiceNo'].n_unique():,}."  ## Muestra la cantidad de facturas encontrada
    )

# Confirma que los controles de integridad fueron correctos
print(
    "Integridad verificada:",  ## Indica que las comprobaciones fueron exitosas
    f"{df.height:,}",  ## Muestra la cantidad de filas encontradas
    "filas y",
    f"{df['InvoiceNo'].n_unique():,}",  ## Muestra la cantidad de facturas diferentes
    "facturas, como se esperaba."  ## Confirma que los valores coinciden con la fuente esperada
)

# Construye una tabla resumen para controlar la calidad de los datos
control = pl.DataFrame({
    "indicador": [
        "filas",  ## Cantidad total de registros
        "facturas",  ## Cantidad de facturas distintas
        "clientes",  ## Cantidad de clientes identificados
        "cancelaciones",  ## Cantidad de registros identificados como cancelaciones
        "cantidades_no_positivas",  ## Registros con cantidades menores o iguales a cero
        "precios_no_positivos"  ## Registros con precios menores o iguales a cero
    ],

    "valor": [
        df.height,  ## Obtiene el número total de filas
        df["InvoiceNo"].n_unique(),  ## Cuenta las facturas diferentes
        df["CustomerID"].drop_nulls().n_unique(),  ## Cuenta clientes excluyendo valores nulos
        int(df["es_cancelacion"].sum()),  ## Suma las cancelaciones identificadas
        int((df["Quantity"] <= 0).sum()),  ## Cuenta cantidades no positivas
        int((df["UnitPrice"] <= 0).sum())  ## Cuenta precios no positivos
    ],
})

# Muestra el resumen de los principales controles de calidad
display(control)

# Obtiene y muestra la primera y última fecha registrada en el dataset
print(
    "Rango:",
    df["InvoiceDate"].min(),
    "->",
    df["InvoiceDate"].max()
)


Fechas que no pudieron parsearse: 0
Integridad verificada: 541,909 filas y 25,900 facturas, como se esperaba.


indicador,valor
str,i64
"""filas""",541909
"""facturas""",25900
"""clientes""",4372
"""cancelaciones""",9288
"""cantidades_no_positivas""",10624
"""precios_no_positivos""",2517


Rango: 2010-12-01 08:26:00 -> 2011-12-09 12:50:00


### **Control de calidad e integridad de los datos**

- **Fechas no parseadas:** 0 → Todas las fechas fueron interpretadas correctamente.
- **Filas:** 541 909 → Coincide con la cantidad esperada.
- **Facturas:** 25 900 → Coincide con la cantidad esperada.
- **Clientes:** 4 372 → Se identificaron 4 372 clientes.
- **Cancelaciones:** 9 288 → Se identificaron 9 288 registros de cancelación.
- **Cantidades no positivas:** 10 624 → Existen registros con cantidades menores o iguales a cero.
- **Precios no positivos:** 2 517 → Existen registros con precios menores o iguales a cero.
- **Rango de fechas:** 01/12/2010 08:26 → 09/12/2011 12:50.

In [4]:
# Filtra únicamente las transacciones que cumplen las condiciones de una compra válida
compras = df.filter(
    (~pl.col("es_cancelacion"))  ## Excluye las facturas identificadas como cancelaciones
    & (pl.col("Quantity") > 0)  ## Conserva cantidades mayores que cero
    & (pl.col("UnitPrice") > 0)  ## Conserva precios unitarios mayores que cero
    & pl.col("InvoiceDate").is_not_null()  ## Excluye registros sin fecha válida
    & pl.col("CustomerID").is_not_null()  ## Excluye registros sin identificación del cliente
)

# Compara el tamaño del dataset original con las compras válidas
print("Filas originales:", df.height)  ## Cantidad total de registros antes del filtrado

print("Filas de compra validas:", compras.height)  ## Cantidad de registros que cumplen las condiciones

# Calcula qué proporción de los registros originales se conserva
print("Proporcion conservada: {:.1%}".format(compras.height / df.height))  ## Muestra el porcentaje conservado

Filas originales: 541909
Filas de compra validas: 397884
Proporcion conservada: 73.4%


### **Conservación de compras válidas**

De las **541 909 filas originales**, se conservaron **397 884 registros** que cumplen las condiciones establecidas para considerarse compras válidas. Esto representa el **73,4 % del total**, mientras que el **26,6 % restante** fue excluido por presentar alguna condición no válida, como cancelaciones, cantidades o precios no positivos, fechas nulas o clientes sin identificación.

**Conclusión:** se cuenta con un conjunto de **397 884 compras válidas** que servirá como base para calcular las métricas y KPI del análisis.

### Reto de calidad
1. ¿Qué sesgo aparece si contamos cancelaciones como ventas?  

Se genera un sesgo de sobreestimación, porque las cancelaciones se contabilizarían como si fueran ventas reales. Esto inflaría los resultados de ventas, ingresos y desempeño comercial, dando una visión incorrecta del crecimiento de la empresa.

2. ¿Por qué no debemos borrar esos registros de la fuente original?

Porque las cancelaciones forman parte de la información real de las transacciones y pueden ser necesarias para analizar el comportamiento de los clientes y la calidad de las ventas. Por ello, debemos conservar la fuente original y, para el análisis de compras válidas, separar o filtrar las cancelaciones sin perder su trazabilidad.


## Paso 4 — Definición del periodo comparable

La fuente empieza el 1/12/2010 y termina el 9/12/2011. Para no comparar meses incompletos con meses completos, el laboratorio informa sobre **enero–noviembre de 2011**.

**Cuidado con el sesgo de ventana.** La tabla de facturas se construye sobre *todo* el historial disponible, y el recorte a la ventana se aplica **después** de determinar la recurrencia. Si se recorta antes, un cliente que compró en diciembre de 2010 y volvió en enero de 2011 aparece como comprador nuevo, y la North Star crece de forma artificial en los primeros meses: en esta base, enero pasa de 570 a 246 compras recurrentes, un 57 % menos, solo por haber filtrado en el orden equivocado.


In [5]:
# Crea una columna con el año y mes de cada compra
compras = compras.with_columns(  ## Agrega una nueva columna al DataFrame
    pl.col("InvoiceDate").dt.strftime("%Y-%m").alias("mes")  ## Extrae año-mes de la fecha y lo guarda como "mes"
)

# Define el inicio y fin de la ventana temporal comparable
INICIO, FIN = pl.datetime(2011, 1, 1), pl.datetime(2011, 12, 1)  ## Analiza el periodo de enero a noviembre de 2011

# Agrupa las líneas de productos para obtener una fila por factura
facturas = (
    compras.group_by(["InvoiceNo", "CustomerID", "mes"])  ## Agrupa por factura, cliente y mes
    .agg([
        pl.col("InvoiceDate").min().alias("fecha_factura"),  ## Obtiene la primera fecha registrada de cada factura
        pl.col("importe_linea").sum().alias("importe_factura"),  ## Suma el importe de todos los productos de la factura
        pl.col("Quantity").sum().alias("unidades"),  ## Calcula el total de unidades compradas
        pl.col("StockCode").count().alias("lineas")  ## Cuenta cuántas líneas de productos tiene la factura
    ])
)

# Selecciona las facturas que pertenecen a la ventana comparable
en_ventana = facturas.filter(
    (pl.col("fecha_factura") >= INICIO) &  ## Incluye facturas desde el 1 de enero de 2011
    (pl.col("fecha_factura") < FIN)  ## Excluye el 1 de diciembre y posteriores
)

# Muestra la cantidad de facturas de todo el historial
print("Facturas en el historial completo :", facturas["InvoiceNo"].n_unique())  ## Cuenta las facturas diferentes del historial

# Muestra la cantidad de facturas dentro de la ventana comparable
print("Facturas en la ventana comparable :", en_ventana["InvoiceNo"].n_unique())  ## Cuenta las facturas del periodo seleccionado

# Muestra la cantidad de clientes presentes en la ventana
print("Clientes en la ventana            :", en_ventana["CustomerID"].n_unique())  ## Cuenta los clientes diferentes del periodo

# Calcula los ingresos totales dentro de la ventana
print("Ingresos en la ventana (GBP)      :", round(en_ventana["importe_factura"].sum(), 2))  ## Suma los importes y redondea a 2 decimales

# Muestra las primeras filas para comprobar la información agrupada
display(facturas.head())  ## Permite revisar el resultado de la agrupación

Facturas en el historial completo : 18532
Facturas en la ventana comparable : 16354
Clientes en la ventana            : 4173
Ingresos en la ventana (GBP)      : 7820501.22


InvoiceNo,CustomerID,mes,fecha_factura,importe_factura,unidades,lineas
str,str,str,datetime[μs],f64,f64,u32
"""557284""","""17377.0""","""2011-06""",2011-06-19 12:29:00,133.07,63.0,12
"""547703""","""12540.0""","""2011-03""",2011-03-24 17:10:00,1140.76,512.0,34
"""555861""","""12990.0""","""2011-06""",2011-06-07 14:27:00,166.4,80.0,4
"""544941""","""12841.0""","""2011-02""",2011-02-25 10:25:00,169.94,120.0,11
"""550918""","""17630.0""","""2011-04""",2011-04-21 14:03:00,335.5,112.0,19


### **Facturación y clientes en la ventana comparable**

La agrupación de las transacciones permitió obtener **18 532 facturas en todo el historial**. De ellas, **16 354 facturas** corresponden a la ventana comparable de enero a noviembre de 2011.

Durante este periodo se identificaron **4 173 clientes** y se generaron **GBP 7 820 501,22 en ingresos**. Además, cada factura quedó resumida con su importe total, unidades y cantidad de líneas de productos, facilitando el cálculo de las métricas posteriores.

**Conclusión:** la ventana comparable concentra la información necesaria para analizar el comportamiento de las compras y determinar posteriormente los clientes recurrentes y la North Star.

## Paso 5 — Recurrencia y North Star

**Regla operativa del laboratorio:** un cliente se considera recurrente desde su segunda factura válida, contada sobre **todo el historial disponible**. Su primera compra no se reclasifica retrospectivamente.

Enero de 2011 conserva un sesgo residual, porque solo dispone de un mes previo de historial. Por eso se marca como **mes de calentamiento** y se excluye de las comparaciones de variación mensual.


In [6]:
# Ordena las facturas por cliente, fecha y número de factura
facturas = facturas.sort(["CustomerID", "fecha_factura", "InvoiceNo"])  ## Permite establecer el orden cronológico de las compras de cada cliente

# Numera las compras de cada cliente comenzando desde 1
facturas = facturas.with_columns(
    (pl.int_range(pl.len()).over("CustomerID") + 1).alias("n_compra_cliente")  ## Asigna un número consecutivo a cada compra del cliente
)

# Identifica como recurrente toda compra realizada a partir de la segunda compra
facturas = facturas.with_columns(
    (pl.col("n_compra_cliente") >= 2).alias("es_recurrente")  ## La primera compra no es recurrente; desde la segunda sí
)

# Recorta el análisis a la ventana comparable después de determinar la recurrencia
facturas = facturas.filter(
    (pl.col("fecha_factura") >= INICIO) &  ## Incluye las facturas desde enero de 2011
    (pl.col("fecha_factura") < FIN)  ## Excluye las facturas desde diciembre de 2011
)

# Define enero como mes de calentamiento porque no tiene un mes previo comparable
MES_CALENTAMIENTO = "2011-01"  ## Se utilizará para evitar comparaciones incompletas en enero

# Calcula las métricas generales de cada mes
mensual = (
    facturas.group_by("mes")  ## Agrupa las facturas por mes
    .agg([
        pl.col("InvoiceNo").n_unique().alias("facturas_validas"),  ## Cuenta las facturas válidas diferentes
        pl.col("CustomerID").n_unique().alias("clientes_activos"),  ## Cuenta los clientes activos diferentes
        pl.col("importe_factura").sum().alias("ingresos")  ## Suma los ingresos de cada mes
    ])
)

# Calcula las métricas correspondientes únicamente a las compras recurrentes
rec = (
    facturas.filter("es_recurrente")  ## Conserva solo las compras realizadas por clientes recurrentes
    .group_by("mes")  ## Agrupa las compras recurrentes por mes
    .agg([
        pl.col("InvoiceNo").n_unique().alias("compras_recurrentes"),  ## Cuenta las facturas recurrentes
        pl.col("CustomerID").n_unique().alias("clientes_recurrentes"),  ## Cuenta los clientes recurrentes
        pl.col("importe_factura").sum().alias("ingresos_recurrentes")  ## Suma los ingresos generados por recurrentes
    ])
)

# Une las métricas generales y recurrentes en una sola tabla
kpi = mensual.join(rec, on="mes", how="left").fill_null(0).sort("mes")  ## Completa con 0 los meses sin compras recurrentes y ordena cronológicamente

# Calcula los KPI relacionados con el comportamiento recurrente
kpi = kpi.with_columns([
    (pl.col("compras_recurrentes") / pl.col("clientes_recurrentes")).alias("frecuencia_recurrente"),  ## Compras recurrentes promedio por cliente recurrente
    (100 * pl.col("ingresos_recurrentes") / pl.col("ingresos")).alias("participacion_ingreso_recurrente_pct"),  ## Porcentaje de ingresos provenientes de clientes recurrentes
])

# Muestra la tabla final de KPI mensuales
display(kpi)  ## Permite revisar las métricas calculadas para cada mes

mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct
str,u32,u32,f64,u32,u32,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535
"""2011-06""",1393,991,661213.69,1151,773,570140.0,1.489004,86.226285
"""2011-07""",1331,949,600091.011,1143,781,532074.68,1.463508,88.665664
"""2011-08""",1280,935,645343.9,1111,778,568263.68,1.428021,88.055947
"""2011-09""",1755,1266,952838.382,1456,996,806684.471,1.461847,84.661207


### **Evolución mensual de clientes e ingresos recurrentes**

- **Enero 2011:** Se registraron **987 facturas válidas**, 741 clientes activos y **GBP 569 445,04** en ingresos. Los clientes recurrentes generaron **GBP 296 614,01**, equivalente al **52,09 %** de los ingresos.
- **Febrero 2011:** La participación del ingreso recurrente aumentó a **67,08 %**, con 412 clientes recurrentes.
- **Marzo 2011:** El ingreso recurrente representó **69,02 %** de los ingresos mensuales.
- **Abril 2011:** La participación recurrente aumentó a **76,15 %**.
- **Mayo 2011:** Los ingresos recurrentes alcanzaron **83,41 %**, mostrando una mayor dependencia de clientes que ya habían comprado anteriormente.
- **Junio 2011:** La participación llegó a **86,23 %**, con 773 clientes recurrentes.
- **Julio 2011:** Se alcanzó una participación de **88,67 %** de ingresos recurrentes.
- **Agosto 2011:** La participación se mantuvo elevada en **88,06 %**.
- **Septiembre 2011:** Los ingresos recurrentes fueron de **GBP 806 684,47**, representando el **84,66 %** de los ingresos.
- **Octubre 2011:** La participación recurrente fue de **85,82 %**, con 1 054 clientes recurrentes.
- **Noviembre 2011:** Se alcanzó el valor más alto, con **90,37 %** de los ingresos provenientes de clientes recurrentes y **1 397 clientes recurrentes**.

### **Comportamiento de los ingresos recurrentes**

Los resultados muestran una **tendencia creciente en la participación de los ingresos provenientes de clientes recurrentes**. En enero representaban el **52,09 %** de los ingresos, mientras que en noviembre alcanzaron el **90,37 %**.

Esto indica que, durante el periodo analizado, los clientes recurrentes fueron adquiriendo una mayor importancia en la generación de ingresos. Además, la cantidad de clientes recurrentes aumentó de **370 en enero a 1 397 en noviembre**.

La **frecuencia recurrente** se mantuvo relativamente estable, entre **1,43 y 1,67 compras por cliente**, alcanzando su máximo en noviembre.

**Conclusión:** el crecimiento observado no depende únicamente de captar nuevos clientes, sino que existe una participación cada vez mayor de clientes que vuelven a comprar. Esto hace que el comportamiento recurrente sea relevante para analizar el valor generado por el negocio.

### North Star propuesta
**Compras válidas de clientes recurrentes por mes.**

Justifique con los cuatro criterios: valor para el cliente, vínculo con valor empresarial, capacidad de influencia del equipo y descomposición en drivers.

**Respuesta:**

Compras válidas de clientes recurrentes por mes es una buena North Star porque permite medir si la empresa está generando valor recurrente y no solamente crecimiento por nuevas ventas.

- Valor para el cliente: representa que los clientes encuentran suficiente valor en la empresa como para volver a comprar.

- Vínculo con el valor empresarial: las compras recurrentes generan ingresos y muestran una relación más estable con los clientes, reduciendo la dependencia de captar nuevos compradores constantemente.

- Capacidad de influencia del equipo: el equipo puede influir mediante acciones como promociones, recomendaciones de productos, comunicación con clientes y estrategias de fidelización para incentivar nuevas compras.

- Descomposición en drivers: puede dividirse en factores como número de clientes recurrentes y frecuencia de compra, permitiendo identificar qué está impulsando el resultado y dónde actuar.


### Reto 1 — ¿Vanidad o acción?
Clasifique: productos totales, clientes activos mensuales, compras recurrentes, ingresos acumulados, tasa de cancelación y países con ventas. Para cada una indique qué decisión permite tomar.

**Respuesta:**

| **Métrica**                    | **Clasificación** | **¿Qué decisión permite tomar?**                                                                                                                                                       |
| ------------------------------ | ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Productos totales**          | Vanidad           | Permite conocer el tamaño del catálogo, pero por sí sola no indica si los productos generan valor. Puede ayudar a decidir si es necesario ampliar o reducir el catálogo.               |
| **Clientes activos mensuales** | Acción            | Permite decidir si se deben reforzar estrategias para atraer, mantener o recuperar clientes activos.                                                                                   |
| **Compras recurrentes**        | Acción            | Permite decidir qué estrategias de fidelización aplicar para aumentar la frecuencia de compra de los clientes recurrentes.                                                             |
| **Ingresos acumulados**        | Vanidad           | Muestra el crecimiento total de los ingresos, pero no permite conocer por sí solo si ese crecimiento es sostenible o proviene de clientes recurrentes.                                 |
| **Tasa de cancelación**        | Acción            | Permite identificar problemas en las ventas y decidir acciones para reducir las cancelaciones y mejorar la calidad de las transacciones.                                               |
| **Países con ventas**          | Vanidad           | Muestra dónde existen ventas, pero por sí sola no indica qué mercados son rentables ni dónde conviene concentrar esfuerzos. Puede servir como punto de partida para analizar mercados. |

Conclusión: las métricas de acción son aquellas que, además de describir lo que ocurre, permiten identificar una situación y orientar una decisión concreta. Por eso, en este caso, las compras recurrentes tienen especial importancia porque están directamente relacionadas con la North Star propuesta.


# ☕ RECESO — 8:30 p. m. a 8:45 p. m.

## Paso 6 — Árbol de métricas

El árbol es una **hipótesis de gestión**, no una demostración causal.

**Objetivo → North Star → drivers → guardrails**

- Objetivo: incrementar valor recurrente sin deteriorar calidad.
- North Star: compras válidas de clientes recurrentes / mes.
- Driver 1: clientes recurrentes activos.
- Driver 2: frecuencia de compra por recurrente.
- Guardrail 1: tasa de cancelación.
- Guardrail 2: concentración de ingresos en Top 10 clientes.


In [7]:
# Reconstruye la North Star a partir de sus componentes principales
kpi = kpi.with_columns(
    (pl.col("clientes_recurrentes") * pl.col("frecuencia_recurrente")).alias("ns_reconstruida")
    ## Clientes recurrentes × frecuencia recurrente = compras recurrentes esperadas
)

# Calcula la diferencia entre las compras recurrentes originales y las reconstruidas
kpi = kpi.with_columns(
    (pl.col("compras_recurrentes") - pl.col("ns_reconstruida")).alias("error_reconstruccion")
    ## Si el resultado es 0, la relación matemática se cumple correctamente
)

# Muestra las variables necesarias para comprobar la relación
display(
    kpi.select([
        "mes",  ## Periodo analizado
        "compras_recurrentes",  ## Valor observado de la North Star
        "clientes_recurrentes",  ## Cantidad de clientes recurrentes
        "frecuencia_recurrente",  ## Compras promedio por cliente recurrente
        "error_reconstruccion"  ## Diferencia entre el valor observado y el reconstruido
    ])
)

mes,compras_recurrentes,clientes_recurrentes,frecuencia_recurrente,error_reconstruccion
str,u32,u32,f64,f64
"""2011-01""",570,370,1.540541,0.0
"""2011-02""",617,412,1.497573,0.0
"""2011-03""",869,563,1.543517,0.0
"""2011-04""",849,587,1.446337,1.1369e-13
"""2011-05""",1271,809,1.571075,0.0
"""2011-06""",1151,773,1.489004,0.0
"""2011-07""",1143,781,1.463508,0.0
"""2011-08""",1111,778,1.428021,0.0
"""2011-09""",1456,996,1.461847,0.0


### **Consistencia de la descomposición de la North Star**

Los resultados muestran que la relación entre **clientes recurrentes × frecuencia recurrente = compras recurrentes** se cumple correctamente en todos los meses analizados.

En **10 de los 11 meses**, el `error_reconstruccion` es exactamente **0.0**. En abril aparece un valor de **1,1369 × 10⁻¹³**, que es prácticamente cero y corresponde a una diferencia mínima causada por la precisión de los cálculos decimales.

Esto confirma que las **compras recurrentes**, utilizadas como North Star, pueden descomponerse matemáticamente en dos componentes: **cantidad de clientes recurrentes** y **frecuencia de compra**.

**Conclusión:** la reconstrucción es consistente y permite utilizar **clientes recurrentes** y **frecuencia recurrente** como drivers de la North Star.

## Paso 7 — Guardrail 1: tasa de cancelación


In [8]:
# Selecciona las transacciones del periodo comparable con información necesaria
base_total = (
    df.filter(
        pl.col("InvoiceDate").is_not_null()  ## Conserva registros con fecha válida
        & pl.col("CustomerID").is_not_null()  ## Conserva registros con cliente identificado
        & (pl.col("InvoiceDate") >= INICIO)  ## Incluye registros desde enero de 2011
        & (pl.col("InvoiceDate") < FIN)  ## Excluye registros desde diciembre de 2011
    )
    .with_columns(
        pl.col("InvoiceDate").dt.strftime("%Y-%m").alias("mes")  ## Crea el periodo año-mes
    )
)

# Obtiene una fila única por factura, cliente y mes
inv_total = base_total.select(
    ["mes", "InvoiceNo", "CustomerID", "es_cancelacion"]
).unique()  ## Evita contar varias veces una misma factura

# Agrupa las facturas por mes para calcular las cancelaciones
cancel = (
    inv_total.group_by("mes")  ## Agrupa los registros según el mes
    .agg([
        pl.col("InvoiceNo").n_unique().alias("facturas_total"),  ## Cuenta todas las facturas diferentes
        pl.col("es_cancelacion").sum().alias("facturas_canceladas")  ## Cuenta las facturas identificadas como canceladas
    ])

    # Calcula el porcentaje de facturas canceladas sobre el total
    .with_columns(
        (
            100 * pl.col("facturas_canceladas") / pl.col("facturas_total")
        ).alias("tasa_cancelacion_pct")  ## Tasa de cancelación expresada en porcentaje
    )

    .sort("mes")  ## Ordena los resultados cronológicamente
)

# Muestra la tasa de cancelación mensual
display(cancel)  ## Presenta el resumen para su análisis

mes,facturas_total,facturas_canceladas,tasa_cancelacion_pct
str,u32,u32,f64
"""2011-01""",1236,249,20.145631
"""2011-02""",1202,204,16.971714
"""2011-03""",1619,298,18.406424
"""2011-04""",1384,235,16.979769
"""2011-05""",1849,294,15.900487
"""2011-06""",1707,314,18.394845
"""2011-07""",1593,262,16.446955
"""2011-08""",1544,263,17.033679
"""2011-09""",2078,322,15.495669


### **Comportamiento de la tasa de cancelación**

La tasa de cancelación presenta una **tendencia general a la baja** durante el periodo analizado. En enero de 2011 fue de **20,15 %**, siendo el valor más alto, mientras que en noviembre disminuyó hasta **13,87 %**, el valor más bajo del periodo.

Aunque existen algunas variaciones entre meses, la reducción progresiva indica que las cancelaciones representan una proporción cada vez menor del total de facturas. Por ejemplo, en mayo fue de **15,90 %**, en septiembre de **15,50 %** y en octubre de **14,76 %**.

Al mismo tiempo, el número total de facturas aumentó de **1 236 en enero a 3 086 en noviembre**, mientras que las facturas canceladas pasaron de **249 a 428**. Esto significa que las cancelaciones aumentaron en cantidad absoluta, pero **crecieron a un ritmo menor que el total de facturas**.

**Conclusión:** la tasa de cancelación muestra una mejora durante 2011, lo que constituye un indicador útil para vigilar la calidad del crecimiento y evitar que el aumento de ventas esté acompañado por un nivel excesivo de cancelaciones.

## Paso 8 — Guardrail 2: concentración Top 10 clientes


In [9]:
# Calcula el ingreso total generado por cada cliente en cada mes
cliente_mes = (
    facturas.group_by(["mes", "CustomerID"])  ## Agrupa por mes y cliente
    .agg(
        pl.col("importe_factura").sum().alias("ingreso_cliente")  ## Suma el ingreso generado por cada cliente
    )
)

# Ordena los clientes de mayor a menor ingreso dentro de cada mes
ordenado = cliente_mes.sort(
    ["mes", "ingreso_cliente"],  ## Ordena primero por mes y luego por ingreso
    descending=[False, True]  ## Mes ascendente e ingresos descendentes
)

# Asigna una posición a cada cliente según su ingreso mensual
ordenado = ordenado.with_columns(
    (pl.int_range(pl.len()).over("mes") + 1).alias("ranking")  ## El cliente con mayor ingreso recibe el ranking 1
)

# Calcula cuánto representan los 10 clientes con mayores ingresos de cada mes
conc = (
    ordenado.group_by("mes")  ## Agrupa nuevamente los resultados por mes
    .agg([
        pl.col("ingreso_cliente").sum().alias("ingreso_mes"),  ## Calcula el ingreso total del mes

        pl.col("ingreso_cliente")
          .filter(pl.col("ranking") <= 10)  ## Selecciona únicamente los 10 clientes con mayor ingreso
          .sum()
          .alias("ingreso_top10")  ## Suma los ingresos generados por esos 10 clientes
    ])

    # Calcula el porcentaje del ingreso mensual concentrado en los 10 principales clientes
    .with_columns(
        (
            100 * pl.col("ingreso_top10") / pl.col("ingreso_mes")
        ).alias("concentracion_top10_pct")  ## Expresa la concentración como porcentaje
    )

    .sort("mes")  ## Ordena los resultados cronológicamente
)

# Muestra la concentración de ingresos de los principales clientes
display(conc)  ## Permite analizar la dependencia de los 10 clientes principales

mes,ingreso_mes,ingreso_top10,concentracion_top10_pct
str,f64,f64,f64
"""2011-01""",569445.04,193292.89,33.944082
"""2011-02""",447137.35,89670.06,20.054254
"""2011-03""",595500.76,113585.39,19.073929
"""2011-04""",469200.361,73375.7,15.638458
"""2011-05""",678594.56,128119.53,18.880129
"""2011-06""",661213.69,193184.16,29.2166
"""2011-07""",600091.011,121223.38,20.200833
"""2011-08""",645343.9,160995.32,24.947213
"""2011-09""",952838.382,233643.58,24.520799


### **Concentración de ingresos en los 10 principales clientes**

Los resultados muestran que la concentración de ingresos en los **10 clientes principales varía considerablemente** durante el periodo. En enero alcanza su nivel más alto, con **33,94 %**, lo que significa que aproximadamente un tercio de los ingresos del mes provino de solo 10 clientes.

A partir de febrero, la concentración disminuye hasta **20,05 %** y continúa variando durante el año. En junio vuelve a aumentar hasta **29,22 %**, mientras que en noviembre alcanza uno de sus niveles más bajos, con **15,66 %**, a pesar de que los ingresos totales fueron los más altos del periodo (**GBP 1 161 800 aproximadamente**).

Esto indica que el crecimiento de los ingresos hacia noviembre **no depende proporcionalmente de los 10 clientes principales**, sino que los ingresos están más distribuidos entre otros clientes.

**Conclusión:** la concentración Top 10 funciona como un **guardrail**, porque permite vigilar el riesgo de depender demasiado de pocos clientes. En este caso, la reducción de **33,94 % en enero a 15,66 % en noviembre** sugiere una menor dependencia relativa de los principales clientes.

In [10]:
# Integra los guardrails al tablero principal de KPI
kpi = (
    kpi
    .join(
        cancel.select(["mes", "tasa_cancelacion_pct"]),  ## Agrega la tasa de cancelación por mes
        on="mes",  ## Utiliza el mes como clave para unir las tablas
        how="left"  ## Conserva todos los meses del KPI principal
    )
    .join(
        conc.select(["mes", "concentracion_top10_pct"]),  ## Agrega la concentración de ingresos del Top 10
        on="mes",  ## Une nuevamente utilizando el mes
        how="left"  ## Mantiene todos los registros de la tabla principal
    )
    .sort("mes")  ## Ordena los resultados cronológicamente
)

# Calcula la variación mensual de la North Star y de sus principales drivers
kpi = kpi.with_columns([
    (
        100 * (
            pl.col("compras_recurrentes") /
            pl.col("compras_recurrentes").shift(1) - 1
        )
    ).alias("var_ns_pct"),  ## Mide cuánto aumentaron o disminuyeron las compras recurrentes frente al mes anterior

    (
        100 * (
            pl.col("clientes_recurrentes") /
            pl.col("clientes_recurrentes").shift(1) - 1
        )
    ).alias("var_clientes_rec_pct"),  ## Mide el cambio porcentual mensual de clientes recurrentes

    (
        100 * (
            pl.col("frecuencia_recurrente") /
            pl.col("frecuencia_recurrente").shift(1) - 1
        )
    ).alias("var_frecuencia_pct"),  ## Mide el cambio porcentual mensual en la frecuencia de compra
])

# Muestra el tablero completo con KPI, drivers y guardrails
display(kpi)  ## Permite analizar la evolución mensual y detectar cambios relevantes


mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct,ns_reconstruida,error_reconstruccion,tasa_cancelacion_pct,concentracion_top10_pct,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct
str,u32,u32,f64,u32,u32,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259,570.0,0.0,20.145631,33.944082,null,null,null
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469,617.0,0.0,16.971714,20.054254,8.245614,11.351351,-2.789133
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604,869.0,0.0,18.406424,19.073929,40.842788,36.650485,3.067901
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289,849.0,1.1369e-13,16.979769,15.638458,-2.301496,4.262877,-6.295983
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535,1271.0,0.0,15.900487,18.880129,49.705536,37.819421,8.624412
"""2011-06""",1393,991,661213.69,1151,773,570140.0,1.489004,86.226285,1151.0,0.0,18.394845,29.2166,-9.441385,-4.449938,-5.223907
"""2011-07""",1331,949,600091.011,1143,781,532074.68,1.463508,88.665664,1143.0,0.0,16.446955,20.200833,-0.695048,1.034929,-1.712256
"""2011-08""",1280,935,645343.9,1111,778,568263.68,1.428021,88.055947,1111.0,0.0,17.033679,24.947213,-2.79965,-0.384123,-2.424841
"""2011-09""",1755,1266,952838.382,1456,996,806684.471,1.461847,84.661207,1456.0,0.0,15.495669,24.520799,31.053105,28.020566,2.368791


### **Evolución de la North Star, drivers y guardrails**

Los resultados muestran que las **compras recurrentes**, nuestra North Star, presentan un crecimiento importante durante el periodo. Pasan de **570 en enero a 2 334 en noviembre**, aunque existen algunas disminuciones puntuales, como en abril (-2,30 %), junio (-9,44 %) y agosto (-2,80 %).

Este comportamiento está explicado principalmente por el crecimiento de los **clientes recurrentes**. Por ejemplo, en noviembre aumentaron **32,54 %** respecto a octubre. Además, la **frecuencia recurrente** aumentó **12,09 %**, alcanzando **1,67 compras por cliente**, su valor más alto del periodo.

Por otro lado, los guardrails muestran una evolución favorable. La **tasa de cancelación** disminuyó de **20,15 % en enero a 13,87 % en noviembre**. La **concentración del Top 10** también terminó en un nivel menor, pasando de **33,94 % a 15,66 %**, lo que indica una menor dependencia relativa de los principales clientes.

La `ns_reconstruida` coincide con las `compras_recurrentes` en todos los meses y el `error_reconstruccion` es **0**, salvo una diferencia prácticamente insignificante en abril debido a la precisión decimal. Esto confirma que la North Star está correctamente descompuesta mediante sus drivers.

**Conclusión:** el crecimiento de la North Star hacia noviembre está asociado principalmente con un mayor número de **clientes recurrentes** y, en ese último mes, también con una mayor **frecuencia de compra**, mientras que los guardrails no muestran un deterioro proporcional. Por ello, el crecimiento observado presenta señales de ser **crecimiento útil y no solamente aparente**.

### Pregunta 2
Identifique el mes con mayor tasa de cancelación y el mes con mayor concentración Top 10. ¿Qué riesgo representa cada uno?

**Respuesta:**

- **Mayor tasa de cancelación:** **enero de 2011**, con **20,15 %**. El riesgo es que una proporción importante de las facturas termine cancelándose, lo que puede afectar la calidad de las ventas y generar una visión sobreestimada del crecimiento.

- **Mayor concentración Top 10:** **enero de 2011**, con **33,94 %**. El riesgo es una **alta dependencia de pocos clientes**, ya que aproximadamente un tercio de los ingresos del mes provino de los 10 clientes principales. Si alguno de ellos deja de comprar, los ingresos podrían verse afectados significativamente.

**Conclusión:** enero presenta simultáneamente los mayores niveles de **cancelación y concentración**, por lo que es el mes que requiere mayor atención desde el punto de vista de los guardrails.

## Paso 9 — Comparación reproducible con Polars y DuckDB


In [11]:
# Selecciona las métricas necesarias para analizar las caídas
ranking_caida = (
    kpi
    .select([
        "mes",  ## Identifica el mes analizado
        "compras_recurrentes",  ## Muestra el valor de la North Star
        "var_ns_pct",  ## Muestra la variación porcentual respecto al mes anterior
        "tasa_cancelacion_pct",  ## Permite revisar el guardrail de cancelaciones
        "concentracion_top10_pct"  ## Permite revisar el guardrail de concentración
    ])
    .sort("var_ns_pct")  ## Ordena de menor a mayor para colocar primero las mayores caídas
)

# Muestra el ranking de meses según la variación de la North Star
ranking_caida


mes,compras_recurrentes,var_ns_pct,tasa_cancelacion_pct,concentracion_top10_pct
str,u32,f64,f64,f64
"""2011-01""",570,null,20.145631,33.944082
"""2011-06""",1151,-9.441385,18.394845,29.2166
"""2011-08""",1111,-2.79965,17.033679,24.947213
"""2011-04""",849,-2.301496,16.979769,15.638458
"""2011-07""",1143,-0.695048,16.446955,20.200833
"""2011-10""",1571,7.898352,14.759169,21.768432
"""2011-02""",617,8.245614,16.971714,20.054254
"""2011-09""",1456,31.053105,15.495669,24.520799
"""2011-03""",869,40.842788,18.406424,19.073929


### **Meses con mayor y menor variación de la North Star**

El ranking muestra que la **mayor caída de compras recurrentes** ocurrió en **junio de 2011**, con una disminución de **9,44 %** respecto a mayo. En este mes también se observa una tasa de cancelación de **18,39 %** y una concentración Top 10 de **29,22 %**, por lo que conviene revisar qué factores pudieron afectar la recurrencia.

La segunda mayor caída ocurrió en **agosto**, con **-2,80 %**, seguida de **abril**, con **-2,30 %**, y **julio**, con **-0,70 %**.

En contraste, los mayores crecimientos se produjeron en **mayo (+49,71 %)**, **noviembre (+48,57 %)** y **marzo (+40,84 %)**. Noviembre destaca especialmente porque, además del fuerte crecimiento de la North Star, presenta la **menor tasa de cancelación (13,87 %)** y una baja concentración Top 10 (**15,66 %**).

El valor `null` de enero es normal porque **no existe un mes anterior dentro del periodo para calcular su variación**.

**Conclusión:** junio es el principal mes para investigar debido a la caída de la North Star. Sin embargo, los datos no permiten afirmar por sí solos que las cancelaciones o la concentración sean la causa de esa caída; estas métricas deben utilizarse como señales para orientar el análisis.

In [12]:
# Ejecuta una consulta SQL directamente sobre el DataFrame de Polars
consulta = duckdb.sql("""
SELECT mes,
       compras_recurrentes,
       ROUND(var_ns_pct, 1) AS var_ns_pct,
       ROUND(tasa_cancelacion_pct, 1) AS cancelacion_pct,
       ROUND(concentracion_top10_pct, 1) AS concentracion_top10_pct
FROM kpi
ORDER BY var_ns_pct ASC NULLS LAST
""").pl()
# Muestra la consulta ordenada
display(consulta)  ## Permite revisar rápidamente los meses con peor y mejor variación

mes,compras_recurrentes,var_ns_pct,cancelacion_pct,concentracion_top10_pct
str,u32,f64,f64,f64
"""2011-06""",1151,-9.4,18.4,29.2
"""2011-08""",1111,-2.8,17.0,24.9
"""2011-04""",849,-2.3,17.0,15.6
"""2011-07""",1143,-0.7,16.4,20.2
"""2011-10""",1571,7.9,14.8,21.8
"""2011-02""",617,8.2,17.0,20.1
"""2011-09""",1456,31.1,15.5,24.5
"""2011-03""",869,40.8,18.4,19.1
"""2011-11""",2334,48.6,13.9,15.7


### **Ranking de la variación mensual de la North Star**

La consulta confirma que **junio de 2011 presenta la mayor caída de la North Star**, con **−9,4 %**, pasando a **1 151 compras recurrentes**. En ese mes, la tasa de cancelación fue de **18,4 %** y la concentración del Top 10 alcanzó **29,2 %**.

Las siguientes caídas fueron **agosto (−2,8 %)**, **abril (−2,3 %)** y **julio (−0,7 %)**. A partir de octubre se observa nuevamente crecimiento, destacando **noviembre (+48,6 %)** y **mayo (+49,7 %)**.

En cuanto a los guardrails, **enero** presenta los valores más altos de cancelación (**20,1 %**) y concentración Top 10 (**33,9 %**), aunque su variación de la North Star aparece como `null` porque no existe un mes anterior para realizar la comparación.

**Conclusión:** el resultado permite identificar rápidamente a **junio como el principal punto de atención**, ya que registra la mayor disminución mensual de las compras recurrentes. Los indicadores de cancelación y concentración sirven como señales adicionales para investigar qué pudo estar relacionado con ese comportamiento, sin afirmar que sean necesariamente la causa.

## Paso 10 — Tablero de decisión


In [13]:
# Crea una figura vacía para construir el gráfico
fig = go.Figure()  ## Permite agregar diferentes series de datos al mismo gráfico

# Agrega la North Star: compras recurrentes por mes
fig.add_trace(
    go.Scatter(
        x=kpi["mes"].to_list(),  ## Utiliza los meses como eje X
        y=kpi["compras_recurrentes"].to_list(),  ## Utiliza las compras recurrentes como eje Y
        mode="lines+markers",  ## Muestra una línea conectada con puntos
        name="North Star"  ## Identifica esta serie como la North Star
    )
)

# Agrega el driver principal: clientes recurrentes por mes
fig.add_trace(
    go.Scatter(
        x=kpi["mes"].to_list(),  ## Utiliza los mismos meses para facilitar la comparación
        y=kpi["clientes_recurrentes"].to_list(),  ## Muestra la cantidad de clientes recurrentes
        mode="lines+markers",  ## Muestra la evolución mediante líneas y puntos
        name="Clientes recurrentes (driver)"  ## Identifica esta serie como driver
    )
)

# Configura el diseño general del gráfico
fig.update_layout(
    title="North Star y driver principal - 2011",  ## Define el título del gráfico
    xaxis_title="Mes",  ## Define el nombre del eje horizontal
    yaxis_title="Cantidad",  ## Define el nombre del eje vertical
    hovermode="x unified"  ## Muestra los valores de ambas series al pasar el cursor por un mismo mes
)

# Muestra el gráfico interactivo
fig.show()  ## Genera el gráfico para analizar visualmente la relación entre North Star y driver


### **Evolución de la North Star y su driver principal**

- La **North Star (compras recurrentes)** presenta una **tendencia general creciente** durante 2011, pasando de **570 compras en enero a 2,334 en noviembre**.

- Los **clientes recurrentes**, que funcionan como driver principal, también aumentan de **370 a 1,397 clientes**, mostrando que el crecimiento de la North Star está acompañado por una mayor cantidad de clientes que vuelven a comprar.

- Se observan **caídas entre mayo y julio**, especialmente en **junio**, donde la North Star disminuye **9.4 %**, siendo la mayor caída mensual del periodo.

- Desde **septiembre** se observa una recuperación clara, alcanzando en noviembre el máximo de **2,334 compras recurrentes**.

- La gráfica permite relacionar el resultado con su driver: cuando aumentan los **clientes recurrentes**, generalmente también aumentan las **compras recurrentes**.

- **Decisión:** se debe priorizar la retención y recuperación de **clientes recurrentes**, ya que su comportamiento influye directamente en la North Star.

In [14]:
fig2 = px.line(kpi.to_pandas(), x="mes", y=["tasa_cancelacion_pct", "concentracion_top10_pct"],
               markers=True, title="Guardrails observados - cancelacion y concentracion")
fig2.update_layout(yaxis_title="Porcentaje (%)", xaxis_title="Mes", legend_title_text="Guardrail")
fig2.show()


### **Evolución de los guardrails durante 2011**

La gráfica muestra el comportamiento de dos guardrails importantes:

- La **tasa de cancelación** presenta una tendencia general a la baja, pasando de **20.1 % en enero a 13.9 % en noviembre**. Esto indica una mejora en la proporción de facturas canceladas.
- La **concentración Top 10** muestra mayor variabilidad. Alcanza su máximo en **enero (33.9 %)** y luego presenta diferentes niveles durante el año.
- En **junio**, la concentración llega a **29.2 %**, mientras que la cancelación alcanza **18.4 %**, por lo que es un mes que requiere mayor atención.
- En **noviembre**, se observa una situación favorable: la cancelación baja a **13.9 %** y la concentración Top 10 a **15.7 %**, mientras la North Star alcanza su máximo.
- **Decisión:** el crecimiento de las compras recurrentes debe mantenerse vigilando estos guardrails para evitar que el aumento de la North Star genere problemas de cancelación o dependencia excesiva de pocos clientes.

In [15]:
# Crea una gráfica de líneas para mostrar la evolución del ingreso recurrente
fig3 = px.line(
    kpi.to_pandas(),  ## Convierte el DataFrame de Polars a Pandas para usar Plotly Express
    x="mes",  ## Define el mes como eje X
    y=["participacion_ingreso_recurrente_pct"],  ## Muestra el porcentaje de ingreso recurrente
    markers=True,  ## Agrega puntos para identificar cada mes
    title="Participacion del ingreso proveniente de compras recurrentes"  ## Define el título de la gráfica
)

# Configura los nombres de los ejes
fig3.update_layout(
    yaxis_title="Porcentaje (%)",  ## El eje Y representa la participación porcentual
    xaxis_title="Mes"  ## El eje X representa los meses analizados
)

# Muestra la gráfica interactiva
fig3.show()  ## Permite observar la evolución mensual del indicador

### **Participación del ingreso proveniente de compras recurrentes**

- La gráfica muestra una **tendencia claramente creciente** durante 2011, pasando de **52.1 % en enero a 90.4 % en noviembre**.

- El mayor crecimiento se observa entre **enero y mayo**, cuando la participación aumenta de **52.1 % a 83.4 %**.

- Entre **junio y julio** continúa aumentando hasta alcanzar aproximadamente **88.7 %**.

- En **agosto y septiembre** se presenta una ligera disminución, llegando a **84.7 % en septiembre**, pero posteriormente vuelve a crecer.

- En **noviembre** se alcanza el máximo del periodo con **90.4 %**, indicando que aproximadamente **9 de cada 10 libras de ingreso** provienen de compras recurrentes.

- **Decisión:** este comportamiento respalda la importancia de la **recurrencia de clientes**, ya que una proporción cada vez mayor de los ingresos está asociada a compradores que regresan.

### Interpretación obligatoria
No basta con decir “la línea bajó”. Responda:
1. ¿Cuánto cambió y en qué periodo?
2. ¿Cómo se comportó el driver principal?
3. ¿Algún guardrail empeoró al mismo tiempo?
4. ¿Qué puede afirmarse como evidencia y qué es solo una inferencia?
5. ¿Quién debería decidir y qué acción ejecutaría?

**Respuesta:**


## Reto 2 — Diagnóstico sin confundir correlación con causalidad


In [16]:
# Identifica el mes con la mayor caída porcentual de la North Star
peor = (
    kpi.filter(
        (pl.col("mes") > MES_CALENTAMIENTO)  ## Excluye enero porque no tiene un mes anterior comparable
        & pl.col("var_ns_pct").is_not_null()  ## Solo considera meses con variación calculada
    )
    .sort("var_ns_pct")  ## Ordena de la caída más negativa a la más positiva
    .head(1)  ## Selecciona únicamente el mes con la mayor caída
)

# Muestra los principales indicadores relacionados con la caída
print("Mayor caida mensual de la North Star")

display(
    peor.select([
        "mes",  ## Mes donde ocurrió la mayor caída
        "var_ns_pct",  ## Porcentaje de variación de la North Star
        "var_clientes_rec_pct",  ## Variación de los clientes recurrentes
        "var_frecuencia_pct",  ## Variación de la frecuencia de compra
        "tasa_cancelacion_pct",  ## Tasa de cancelación del mes
        "concentracion_top10_pct"  ## Porcentaje de ingresos concentrado en los 10 principales clientes
    ])
)


Mayor caida mensual de la North Star


mes,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct,tasa_cancelacion_pct,concentracion_top10_pct
str,f64,f64,f64,f64,f64
"""2011-06""",-9.441385,-4.449938,-5.223907,18.394845,29.2166


### **Mayor caída mensual de la North Star**

- El mes con la **mayor caída de la North Star fue junio de 2011**, con una disminución de **9.4 %** respecto a mayo.

- Esta caída estuvo acompañada por una reducción de **4.4 % en los clientes recurrentes** y de **5.2 % en la frecuencia de compra**, por lo que ambos drivers contribuyeron al descenso.

- La **tasa de cancelación fue de 18.4 %**, mientras que la **concentración Top 10 alcanzó 29.2 %**. Estos valores representan factores que deben vigilarse durante este periodo.

- **Decisión:** junio debe considerarse un mes de atención, priorizando acciones para **recuperar clientes recurrentes y mejorar la frecuencia de compra**, mientras se controlan las cancelaciones y la concentración de ingresos.

A partir de la salida anterior, escriba una conclusión en tres capas:

- **Evidencia:** lo que muestran los datos.
- **Inferencia:** explicación plausible, sin afirmar causalidad.
- **Decisión:** acción concreta que debería evaluar el responsable.

**Respuesta:**
- **Evidencia:** En **junio de 2011**, la North Star presentó su mayor caída mensual, disminuyendo **9.4 %**. Al mismo tiempo, los **clientes recurrentes disminuyeron 4.4 %** y la **frecuencia de compra 5.2 %**. La tasa de cancelación fue de **18.4 %** y la concentración Top 10 alcanzó **29.2 %**.

- **Inferencia:** La disminución de la North Star podría estar relacionada con una **menor cantidad de clientes recurrentes y una menor frecuencia de compra**. Además, el nivel de cancelaciones y concentración de ingresos son señales que deberían investigarse, aunque **los datos no permiten afirmar causalidad**.

- **Decisión:** El responsable debería evaluar acciones para **recuperar clientes recurrentes y aumentar su frecuencia de compra**, revisando al mismo tiempo las causas de las cancelaciones y el nivel de concentración de ingresos durante junio.

## Diccionario de KPI — completar



| KPI | Fórmula / unidad | Fuente / frecuencia | Responsable | Meta / alerta | Acción |
|---|---|---|---|---|---|
| **Compras recurrentes/mes** | Número de compras válidas realizadas por clientes recurrentes / compras por mes | UCI / mensual | Equipo Comercial | **Aumentar** / alerta si disminuye | Implementar acciones de retención y recuperación de clientes |
| **Clientes recurrentes activos** | Número de clientes recurrentes que realizaron al menos una compra en el mes / clientes | UCI / mensual | Equipo Comercial | **Aumentar** / alerta si disminuye | Ejecutar campañas de reactivación y fidelización |
| **Frecuencia recurrente** | Compras recurrentes ÷ clientes recurrentes / compras por cliente | UCI / mensual | Marketing / Comercial | **Aumentar** / alerta si disminuye | Promover nuevas compras mediante ofertas y estrategias de fidelización |
| **Tasa de cancelación** | Facturas canceladas ÷ facturas totales × 100 / porcentaje | UCI / mensual | Operaciones | **Disminuir** / alerta si aumenta | Investigar causas de cancelación y corregir problemas del proceso |
| **Concentración Top 10** | Ingresos de los 10 principales clientes ÷ ingresos totales × 100 / porcentaje | UCI / mensual | Comercial / Gerencia | **Disminuir** / alerta si aumenta | Diversificar la cartera de clientes y reducir la dependencia de pocos clientes |

> Las metas y alertas que proponga son supuestos académicos de gestión, no metas oficiales de la empresa.


---

## Reto de aplicación y retroalimentación

En esta sección se aplicarán los procedimientos desarrollados durante la sesión a nuevas situaciones de análisis. Cada ejercicio requiere modificar, completar o construir código a partir de las tablas ya procesadas. Posteriormente, los resultados obtenidos deberán interpretarse brevemente desde una perspectiva empresarial. El propósito es comprobar la comprensión de las técnicas utilizadas y fortalecer la capacidad de adaptar el análisis ante nuevas preguntas de negocio.

**Indicaciones generales.** Los ejercicios operan sobre `kpi_reto`, copia de trabajo de la tabla de KPI, y sobre `facturas`, ya construida en la Actividad 2. Los datos proceden íntegramente del repositorio UCI; no corresponde generar ni sustituir valores en ningún caso. Cada respuesta escrita no debe exceder cuatro líneas.

**Duración en sesión:** 25 minutos. Los ejercicios que no concluyan se completan como avance del entregable.


In [23]:
# Crea una copia independiente de la tabla de KPI para trabajar en los retos
kpi_reto = kpi.clone()  ## Permite modificar la copia sin alterar la tabla kpi original

# Verifica las dimensiones de la copia creada
print("Copia de trabajo creada:", kpi_reto.shape)  ## Muestra filas y columnas disponibles

# Muestra las columnas disponibles para los siguientes ejercicios
print("Columnas disponibles:", kpi_reto.columns)  ## Permite identificar las variables que podemos utilizar

# Muestra las primeras filas para comprobar que la copia contiene los datos esperados
display(kpi_reto.head())  ## Verifica visualmente la estructura de la tabla


Copia de trabajo creada: (11, 16)
Columnas disponibles: ['mes', 'facturas_validas', 'clientes_activos', 'ingresos', 'compras_recurrentes', 'clientes_recurrentes', 'ingresos_recurrentes', 'frecuencia_recurrente', 'participacion_ingreso_recurrente_pct', 'ns_reconstruida', 'error_reconstruccion', 'tasa_cancelacion_pct', 'concentracion_top10_pct', 'var_ns_pct', 'var_clientes_rec_pct', 'var_frecuencia_pct']


mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct,ns_reconstruida,error_reconstruccion,tasa_cancelacion_pct,concentracion_top10_pct,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct
str,u32,u32,f64,u32,u32,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259,570.0,0.0,20.145631,33.944082,null,null,null
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469,617.0,0.0,16.971714,20.054254,8.245614,11.351351,-2.789133
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604,869.0,0.0,18.406424,19.073929,40.842788,36.650485,3.067901
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289,849.0,1.1369e-13,16.979769,15.638458,-2.301496,4.262877,-6.295983
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535,1271.0,0.0,15.900487,18.880129,49.705536,37.819421,8.624412


### **Verificación de la tabla de KPI para los retos**

- La copia de trabajo **`kpi_reto`** contiene **11 meses y 16 columnas**, por lo que conserva toda la información necesaria para los siguientes ejercicios.
- Incluye la **North Star**, sus **drivers**, los **guardrails**, las variaciones mensuales y el **error de reconstrucción**.
- Los valores observados mantienen consistencia con el análisis anterior; por ejemplo, en mayo se registran **1 271 compras recurrentes** y **809 clientes recurrentes**.
- **Conclusión:** `kpi_reto` está correctamente preparada como tabla de trabajo para continuar con los retos de análisis.

### Ejercicio 1 — Construcción de una tasa de recurrencia mensual

La North Star del laboratorio se expresa en cantidad de compras recurrentes, magnitud absoluta que crece cuando aumenta la actividad total del negocio. Una magnitud absoluta, sin embargo, no permite distinguir si la recurrencia mejora o si simplemente hay más transacciones de cualquier tipo.

La tasa de recurrencia corrige esa limitación: expresa qué proporción de las facturas del mes corresponde a clientes que ya habían comprado con anterioridad. Se trata de una magnitud relativa y, por tanto, comparable entre meses de distinto volumen.

**Se solicita:**

1. Calcular, para cada mes, la cantidad total de facturas y la cantidad de facturas recurrentes a partir de `facturas`.
2. Construir la tasa de recurrencia mensual expresada en porcentaje e incorporarla a `kpi_reto`.
3. Identificar el mes con la tasa más alta y el mes con la tasa más baja.
4. Comparar el ordenamiento por tasa de recurrencia con el ordenamiento por compras recurrentes.

**Tiempo estimado:** 5 minutos.


### **Cálculo de la tasa de recurrencia mensual**

$$
\text{Tasa de recurrencia} = \frac{\text{Facturas recurrentes}}{\text{Facturas totales}} \times 100
$$


La tasa indica qué **porcentaje de las facturas del mes corresponde a clientes que ya habían comprado anteriormente**.

In [24]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 1
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

# Calcula las facturas totales y recurrentes por cada mes
recurrencia = (
    facturas.group_by("mes")  ## Agrupa las facturas según el mes
    .agg([
        pl.col("InvoiceNo").n_unique().alias("facturas_totales"),  ## Cuenta las facturas válidas del mes
        pl.col("InvoiceNo").filter(pl.col("es_recurrente")).n_unique().alias("facturas_recurrentes")  ## Cuenta las facturas de clientes recurrentes
    ])
    .with_columns(
        (100 * pl.col("facturas_recurrentes") / pl.col("facturas_totales"))
        .alias("tasa_recurrencia_pct")  ## Calcula qué porcentaje de facturas corresponde a clientes recurrentes
    )
    .sort("mes")  ## Ordena los resultados cronológicamente
)

# Incorpora la tasa de recurrencia a la tabla de KPI del reto
kpi_reto = kpi_reto.join(
    recurrencia.select(["mes", "tasa_recurrencia_pct"]),  ## Selecciona únicamente las columnas necesarias
    on="mes",  ## Utiliza el mes como clave de unión
    how="left"  ## Conserva todos los meses de kpi_reto
)

# Muestra la tasa de recurrencia mensual
display(
    recurrencia  ## Presenta las facturas totales, recurrentes y la tasa calculada
)

# Identifica el mes con la tasa de recurrencia más alta
mayor_recurrencia = recurrencia.sort(
    "tasa_recurrencia_pct", descending=True  ## Ordena de mayor a menor tasa
).head(1)

# Identifica el mes con la tasa de recurrencia más baja
menor_recurrencia = recurrencia.sort(
    "tasa_recurrencia_pct"  ## Ordena de menor a mayor tasa
).head(1)

print("Mes con mayor tasa de recurrencia:")
display(mayor_recurrencia)

print("Mes con menor tasa de recurrencia:")
display(menor_recurrencia)

# Compara el ranking por tasa de recurrencia con el ranking por compras recurrentes
comparacion = recurrencia.select([
    "mes",
    "tasa_recurrencia_pct",
    "facturas_recurrentes"
]).sort("tasa_recurrencia_pct", descending=True)  ## Ordena los meses según la tasa de recurrencia

display(comparacion)


mes,facturas_totales,facturas_recurrentes,tasa_recurrencia_pct
str,u32,u32,f64
"""2011-01""",987,570,57.75076
"""2011-02""",997,617,61.885657
"""2011-03""",1321,869,65.783497
"""2011-04""",1149,849,73.890339
"""2011-05""",1555,1271,81.736334
"""2011-06""",1393,1151,82.627423
"""2011-07""",1331,1143,85.875282
"""2011-08""",1280,1111,86.796875
"""2011-09""",1755,1456,82.962963


Mes con mayor tasa de recurrencia:


mes,facturas_totales,facturas_recurrentes,tasa_recurrencia_pct
str,u32,u32,f64
"""2011-11""",2657,2334,87.843432


Mes con menor tasa de recurrencia:


mes,facturas_totales,facturas_recurrentes,tasa_recurrencia_pct
str,u32,u32,f64
"""2011-01""",987,570,57.75076


mes,tasa_recurrencia_pct,facturas_recurrentes
str,f64,u32
"""2011-11""",87.843432,2334
"""2011-08""",86.796875,1111
"""2011-07""",85.875282,1143
"""2011-09""",82.962963,1456
"""2011-06""",82.627423,1151
"""2011-05""",81.736334,1271
"""2011-10""",81.441161,1571
"""2011-04""",73.890339,849
"""2011-03""",65.783497,869


### **Comportamiento de la tasa de recurrencia mensual**

- La **tasa de recurrencia aumenta** de **57.8 % en enero a 87.8 % en noviembre**, mostrando una mayor proporción de facturas asociadas a clientes recurrentes.
- El mes con la **mayor tasa de recurrencia fue noviembre**, con **87.8 %**, correspondiente a **2 334 facturas recurrentes de 2 657 facturas totales**.
- El mes con la **menor tasa de recurrencia fue enero**, con **57.8 %**, correspondiente a **570 facturas recurrentes de 987 facturas totales**.
- El ordenamiento muestra que **noviembre también presenta el mayor número de compras recurrentes**, mientras que enero presenta el menor valor, aunque la tasa permite comparar los meses considerando su volumen total de facturas.

**Pregunta 3.** ¿Coincide el mes con mayor cantidad de compras recurrentes con el mes de mayor tasa de recurrencia? Explique qué implicancia tiene esa diferencia para la medición del negocio.

**Respuesta (máximo cuatro líneas):**

### **Diferencia entre volumen y proporción de recurrencia**

**Respuesta:** No, ambos meses coinciden: **noviembre** presenta tanto la mayor cantidad de compras recurrentes (**2 334**) como la mayor tasa de recurrencia (**87.8 %**). Esto indica que en noviembre hubo un alto volumen y, además, una mayor proporción de facturas recurrentes. La tasa permite comparar el desempeño entre meses sin depender únicamente del tamaño del negocio. **Por ello, ambas métricas se complementan para evaluar el crecimiento y la recurrencia de los clientes.**


### Ejercicio 2 — Incorporación del ticket promedio como driver económico

El árbol de métricas descompuso la North Star en clientes recurrentes y frecuencia de compra. Ninguno de esos dos drivers recoge el valor económico de cada transacción: dos meses con idéntica cantidad de compras recurrentes pueden presentar ingresos muy distintos si el importe promedio de la factura cambia.

El ticket promedio de la factura recurrente completa esa descripción y permite establecer si el crecimiento observado proviene de mayor actividad o de mayor valor por transacción.

**Se solicita:**

1. Calcular, a partir de `facturas`, el importe promedio de las facturas recurrentes de cada mes.
2. Incorporar el resultado a `kpi_reto` en la columna `ticket_promedio_recurrente`.
3. Calcular su variación mensual porcentual, siguiendo el mismo procedimiento empleado para los demás drivers.
4. Contrastar los meses de mayor caída de la North Star con el comportamiento del ticket promedio en esos mismos meses.

**Tiempo estimado:** 5 minutos.


### **Cálculo del ticket promedio recurrente**

$$
\text{Ticket promedio recurrente} =
\frac{\text{Ingresos recurrentes}}
{\text{Facturas recurrentes}}
$$

In [25]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 2
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------
# Calcula los ingresos y las facturas recurrentes de cada mes
ticket_recurrente = (
    facturas.filter(
        pl.col("es_recurrente")  ## Considera únicamente las facturas de clientes recurrentes
    )
    .group_by("mes")  ## Agrupa las facturas recurrentes por mes
    .agg([
        pl.col("importe_factura").sum().alias("ingresos_recurrentes"),  ## Suma los ingresos generados por facturas recurrentes
        pl.col("InvoiceNo").n_unique().alias("facturas_recurrentes")  ## Cuenta las facturas recurrentes del mes
    ])
    .with_columns(
        (
            pl.col("ingresos_recurrentes") / pl.col("facturas_recurrentes")
        ).alias("ticket_promedio_recurrente")  ## Calcula el ingreso promedio generado por cada factura recurrente
    )
    .sort("mes")  ## Ordena los meses cronológicamente
)

# Incorpora el ticket promedio recurrente a la tabla de KPI del reto
kpi_reto = kpi_reto.drop("ticket_promedio_recurrente") if "ticket_promedio_recurrente" in kpi_reto.columns else kpi_reto  ## Evita duplicar la columna si se ejecuta nuevamente

kpi_reto = kpi_reto.join(
    ticket_recurrente.select(["mes", "ticket_promedio_recurrente"]),  ## Selecciona el mes y el ticket calculado
    on="mes",  ## Utiliza el mes como clave de unión
    how="left"  ## Conserva todos los meses de kpi_reto
)

# Calcula la variación mensual porcentual del ticket promedio recurrente
kpi_reto = kpi_reto.with_columns(
    (
        100 * (
            pl.col("ticket_promedio_recurrente") /
            pl.col("ticket_promedio_recurrente").shift(1) - 1
        )
    ).alias("var_ticket_promedio_pct")  ## Mide cuánto aumenta o disminuye el ticket respecto al mes anterior
)

# Muestra el ticket promedio y su variación mensual
display(
    kpi_reto.select([
        "mes",  ## Mes analizado
        "compras_recurrentes",  ## North Star del laboratorio
        "ticket_promedio_recurrente",  ## Valor promedio de cada factura recurrente
        "var_ticket_promedio_pct"  ## Variación porcentual respecto al mes anterior
    ])
)

mes,compras_recurrentes,ticket_promedio_recurrente,var_ticket_promedio_pct
str,u32,f64,f64
"""2011-01""",570,520.375456,null
"""2011-02""",617,486.122139,-6.582424
"""2011-03""",869,472.992094,-2.700977
"""2011-04""",849,420.817397,-11.030776
"""2011-05""",1271,445.34989,5.829724
"""2011-06""",1151,495.34318,11.225621
"""2011-07""",1143,465.507157,-6.023304
"""2011-08""",1111,511.488461,9.877679
"""2011-09""",1456,554.041532,8.319459


### **Evolución del ticket promedio de las compras recurrentes**

- El **ticket promedio recurrente** varía durante el periodo, pasando de **GBP 520.38 en enero a GBP 449.86 en noviembre**.
- La mayor caída se presenta en **noviembre**, con una disminución de **20.8 %** respecto a octubre, aunque las compras recurrentes aumentaron hasta **2 334**.
- En **junio**, mes con la mayor caída de la North Star, el ticket promedio **aumentó 11.2 %**, pasando de **GBP 445.35 a GBP 495.34**.
- **Conclusión:** la caída de la North Star en junio no estuvo acompañada por una disminución del ticket promedio; por ello, el comportamiento observado sugiere que la caída estuvo más relacionada con los **clientes recurrentes y la frecuencia de compra** que con el valor promedio de cada factura.

**Pregunta 4.** En el mes de mayor caída de la North Star, ¿el ticket promedio recurrente aumentó o disminuyó? ¿Qué sugiere ese comportamiento conjunto?

**Respuesta (máximo cuatro líneas):**

---
### **Relación entre la caída de la North Star y el ticket promedio**

**Respuesta:** En **junio de 2011**, el ticket promedio recurrente **aumentó 11.2 %**, mientras la North Star disminuyó **9.4 %**. Esto sugiere que la caída estuvo más relacionada con una **menor cantidad de clientes recurrentes y frecuencia de compra**, y no con una reducción del valor promedio por factura. **Por ello, el problema principal a investigar estaría en la actividad de los clientes y no en el valor de cada transacción.**

### Ejercicio 3 — Definición de un guardrail de dependencia del cliente principal

La Actividad 2 incorporó un guardrail de concentración sobre los diez principales clientes. Ese umbral describe la dependencia agregada del negocio, pero no revela si dicha concentración se explica por un único cliente de gran tamaño, situación que constituye un riesgo operativo de naturaleza distinta.

Un guardrail de dependencia del cliente principal mide qué proporción del ingreso mensual corresponde al cliente de mayor facturación. Su finalidad no es optimizarse, sino advertir cuando la concentración alcanza un nivel que compromete la continuidad del negocio.

**Se solicita:**

1. Calcular, para cada mes, el ingreso del cliente de mayor facturación. La tabla `ordenado` ya dispone de la columna `ranking` calculada dentro de cada mes.
2. Expresar ese ingreso como porcentaje del ingreso total del mes e incorporarlo a `kpi_reto`.
3. Declarar explícitamente un umbral de alerta como criterio pedagógico del ejercicio, no como norma sectorial.
4. Identificar los meses que superan dicho umbral.

**Tiempo estimado:** 5 minutos.


### **Cálculo de la dependencia del cliente principal**

$$
\text{Dependencia del cliente principal} =
\frac{\text{Ingreso del cliente principal}}
{\text{Ingreso total del mes}}
\times 100
$$

In [26]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 3
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------
# Define el umbral pedagógico de alerta para la dependencia del cliente principal
UMBRAL_DEPENDENCIA = 20.0  ## Se considera alerta cuando un solo cliente representa más del 20 % del ingreso mensual

# Obtiene el ingreso del cliente con mayor facturación de cada mes
cliente_principal = (
    ordenado.filter(
        pl.col("ranking") == 1  ## Selecciona únicamente al cliente con mayor ingreso de cada mes
    )
    .select([
        "mes",  ## Identifica el mes analizado
        pl.col("ingreso_cliente").alias("ingreso_cliente_principal")  ## Guarda el ingreso del cliente principal
    ])
)

# Calcula el porcentaje que representa el cliente principal sobre el ingreso total del mes
dependencia = (
    cliente_principal
    .join(
        kpi_reto.select(["mes", "ingresos"]),  ## Obtiene el ingreso total de cada mes
        on="mes",  ## Une ambas tablas mediante el mes
        how="left"  ## Conserva todos los meses con cliente principal
    )
    .with_columns(
        (
            100 * pl.col("ingreso_cliente_principal") / pl.col("ingresos")
        ).alias("dependencia_cliente_principal_pct")  ## Calcula la participación del cliente principal
    )
    .sort("mes")  ## Ordena los resultados cronológicamente
)

# Incorpora el nuevo guardrail a la tabla de KPI del reto
kpi_reto = kpi_reto.drop("dependencia_cliente_principal_pct") if "dependencia_cliente_principal_pct" in kpi_reto.columns else kpi_reto  ## Evita duplicar la columna al ejecutar nuevamente

kpi_reto = kpi_reto.join(
    dependencia.select(["mes", "dependencia_cliente_principal_pct"]),  ## Selecciona el mes y el guardrail calculado
    on="mes",  ## Utiliza el mes como clave
    how="left"  ## Conserva todos los meses de la tabla de KPI
)

# Identifica los meses que superan el umbral de alerta
meses_alerta = dependencia.filter(
    pl.col("dependencia_cliente_principal_pct") > UMBRAL_DEPENDENCIA  ## Selecciona los meses donde la dependencia supera el 20 %
).sort(
    "dependencia_cliente_principal_pct", descending=True  ## Ordena de mayor a menor dependencia
)

# Muestra el resultado mensual del nuevo guardrail
display(
    dependencia.select([
        "mes",  ## Mes analizado
        "ingreso_cliente_principal",  ## Ingreso generado por el cliente principal
        "ingresos",  ## Ingreso total del mes
        "dependencia_cliente_principal_pct"  ## Porcentaje de dependencia
    ])
)

# Muestra el umbral y los meses que lo superan
print(f"Umbral pedagógico de alerta: {UMBRAL_DEPENDENCIA:.1f}%")
print("Meses que superan el umbral:")
display(meses_alerta)

mes,ingreso_cliente_principal,ingresos,dependencia_cliente_principal_pct
str,f64,f64,f64
"""2011-01""",77183.6,569445.04,13.554179
"""2011-02""",22797.46,447137.35,5.098536
"""2011-03""",21462.4,595500.76,3.604093
"""2011-04""",21535.9,469200.361,4.589915
"""2011-05""",28408.14,678594.56,4.18632
"""2011-06""",41959.44,661213.69,6.345821
"""2011-07""",26464.99,600091.011,4.410163
"""2011-08""",40327.81,645343.9,6.249042
"""2011-09""",75412.64,952838.382,7.914526


Umbral pedagógico de alerta: 20.0%
Meses que superan el umbral:


mes,ingreso_cliente_principal,ingresos,dependencia_cliente_principal_pct
str,f64,f64,f64


### **Dependencia del cliente principal y nivel de alerta**

- La **mayor dependencia del cliente principal** se presentó en **enero de 2011**, con **13.6 %** de los ingresos mensuales.
- Ningún mes superó el **umbral pedagógico de alerta del 20 %**, ya que todas las tasas se mantuvieron por debajo de dicho límite.
- La dependencia disminuyó hasta **2.4 % en noviembre**, indicando una menor concentración de los ingresos en un único cliente.
- **Conclusión:** no se identifican meses en situación de alerta según el criterio establecido, por lo que el negocio **no presenta una dependencia elevada de un solo cliente** durante el periodo analizado.

**Pregunta 5.** ¿En cuántos meses el cliente principal supera el umbral declarado y qué decisión empresarial justificaría ese resultado?

**Respuesta (máximo cuatro líneas):**

---
### **Dependencia del cliente principal frente al umbral de alerta**

**Respuesta:** El cliente principal **no supera el umbral declarado en ningún mes**, por lo que se registran **0 meses en alerta**. La mayor dependencia fue de **13.6 % en enero**, por debajo del umbral del **20 %**. La decisión empresarial sería **mantener el monitoreo de la concentración de ingresos** para prevenir una dependencia excesiva de un solo cliente.


### Ejercicio 4 — Consulta de meses saludables con condiciones múltiples en DuckDB

En la Actividad 2 se empleó DuckDB para ordenar los meses según la variación de la North Star. Una consulta orientada a la decisión, sin embargo, no se limita a ordenar: delimita el subconjunto de periodos que satisfacen simultáneamente el objetivo de crecimiento y las restricciones fijadas por los guardrails.

Un mes de crecimiento acompañado de un deterioro en la tasa de cancelación no constituye un mes saludable. Esta consulta materializa esa distinción en una regla reproducible.

**Se solicita:**

1. Construir sobre `kpi_reto` una consulta SQL que incluya `SELECT`, `WHERE`, condiciones múltiples enlazadas con `AND` y `ORDER BY`.
2. Establecer como criterios una variación de la North Star positiva y una tasa de cancelación inferior al promedio del periodo.
3. Ordenar el resultado por variación de la North Star de mayor a menor.
4. Determinar cuántos meses satisfacen ambos criterios.

**Tiempo estimado:** 5 minutos.


### **Criterios para identificar un mes saludable**

$$
\text{Mes saludable} =
(\text{Variación North Star} > 0)
\land
(\text{Tasa de cancelación} < \text{Promedio de cancelación})
$$

In [31]:
# Calcula el promedio de la tasa de cancelación del periodo
promedio_cancelacion = duckdb.sql("""
SELECT AVG(tasa_cancelacion_pct) AS promedio_cancelacion
FROM kpi_reto
""").fetchone()[0]  ## Obtiene el promedio de cancelación para utilizarlo como referencia

print(f"Promedio de cancelación del periodo: {promedio_cancelacion:.2f}%")  ## Muestra el promedio calculado

# Busca los meses que cumplen simultáneamente las condiciones de crecimiento y calidad
meses_saludables = duckdb.sql("""
SELECT
    mes,
    compras_recurrentes,
    ROUND(var_ns_pct, 1) AS var_ns_pct,
    ROUND(tasa_cancelacion_pct, 1) AS tasa_cancelacion_pct
FROM kpi_reto
WHERE var_ns_pct > 0
  AND tasa_cancelacion_pct < (
      SELECT AVG(tasa_cancelacion_pct)
      FROM kpi_reto
  )
ORDER BY var_ns_pct DESC
""").pl()  ## Convierte el resultado de DuckDB nuevamente a Polars

# Muestra los meses que cumplen ambos criterios
display(meses_saludables)  ## Permite revisar los meses considerados saludables

# Cuenta la cantidad de meses saludables
print("Cantidad de meses saludables:", meses_saludables.height)  ## Muestra cuántos meses cumplen las condiciones


Promedio de cancelación del periodo: 16.76%


mes,compras_recurrentes,var_ns_pct,tasa_cancelacion_pct
str,u32,f64,f64
"""2011-05""",1271,49.7,15.9
"""2011-11""",2334,48.6,13.9
"""2011-09""",1456,31.1,15.5
"""2011-10""",1571,7.9,14.8


Cantidad de meses saludables: 4


### **Identificación de meses con crecimiento saludable**

- El **promedio de cancelación del periodo fue de 16.76 %**.
- Se identificaron **4 meses saludables**: **mayo, septiembre, octubre y noviembre**, porque presentan crecimiento positivo de la North Star y una tasa de cancelación inferior al promedio.
- **Mayo** tuvo el mayor crecimiento de la North Star (**49.7 %**) con una cancelación de **15.9 %**, mientras que **noviembre** alcanzó **2 334 compras recurrentes** y una cancelación de **13.9 %**.
- **Conclusión:** estos meses muestran un crecimiento de la North Star acompañado de un nivel de cancelación inferior al promedio, por lo que representan periodos de **crecimiento saludable** según los criterios establecidos.

**Pregunta 6.** ¿Cuántos meses del periodo pueden calificarse como saludables según los criterios establecidos y qué sugiere esa proporción sobre la solidez del crecimiento?

**Respuesta (máximo cuatro líneas):**

---
### **Proporción de meses con crecimiento saludable**

**Respuesta:** Se identificaron **4 de los 11 meses** como saludables, lo que representa aproximadamente el **36.4 % del periodo**. Esto sugiere que el crecimiento no fue consistentemente saludable durante todo el periodo analizado. Por ello, sería importante investigar los meses restantes para identificar qué factores limitaron la solidez del crecimiento.

### Ejercicio 5 — Representación de la relación entre el driver y la North Star

El tablero construido en la Actividad 2 presenta la North Star y su driver como series temporales paralelas. Esa disposición permite observar la evolución de ambas magnitudes, pero no muestra con claridad si sus variaciones se corresponden entre sí.

Un gráfico de dispersión que enfrente la variación del driver con la variación de la North Star hace visible esa correspondencia: los puntos alineados sobre una tendencia ascendente indican que el driver acompaña el movimiento del resultado, mientras que los puntos dispersos advierten que otros factores intervienen. Corresponde recordar que la correspondencia observada describe una asociación y no acredita una relación causal.

**Se solicita:**

1. Construir con Plotly un gráfico de dispersión que sitúe la variación porcentual de clientes recurrentes en el eje horizontal y la variación porcentual de la North Star en el eje vertical.
2. Identificar cada punto con el mes correspondiente.
3. Rotular los ejes y titular el gráfico de modo que resulte interpretable sin recurrir al código.
4. Excluir de manera explícita el mes de calentamiento, que carece de variación calculable.

**Tiempo estimado:** 5 minutos.


In [33]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 5
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------
# Filtra los meses con variaciones calculables
dispersion = kpi_reto.filter(
    (pl.col("mes") > MES_CALENTAMIENTO)  ## Excluye enero, que no tiene variación mensual
    & pl.col("var_clientes_rec_pct").is_not_null()  ## Conserva solo meses con variación de clientes recurrentes
    & pl.col("var_ns_pct").is_not_null()  ## Conserva solo meses con variación de la North Star
)

# Convierte los datos a Pandas para utilizarlos con Plotly
dispersion_pd = dispersion.to_pandas()  ## Prepara la tabla para construir el gráfico

# Construye el gráfico de dispersión
fig = px.scatter(
    dispersion_pd,
    x="var_clientes_rec_pct",  ## Coloca la variación de clientes recurrentes en el eje horizontal
    y="var_ns_pct",  ## Coloca la variación de la North Star en el eje vertical
    text="mes",  ## Identifica cada punto con el mes correspondiente
    title="Relación entre la variación de clientes recurrentes y la North Star"
)

# Personaliza las etiquetas para facilitar la interpretación
fig.update_traces(
    textposition="top center"  ## Coloca el nombre del mes encima de cada punto
)

fig.update_layout(
    xaxis_title="Variación de clientes recurrentes (%)",  ## Etiqueta del eje X
    yaxis_title="Variación de la North Star (%)",  ## Etiqueta del eje Y
    hovermode="closest"  ## Muestra información del punto más cercano al pasar el cursor
)

# Muestra el gráfico
fig.show()

### **Relación entre clientes recurrentes y la North Star**

- Se observa una **relación positiva**: los meses con mayor crecimiento de clientes recurrentes generalmente presentan también mayor crecimiento de la North Star.
- **Mayo** y **noviembre** destacan con crecimientos de la North Star cercanos al **50 %**, acompañados por aumentos de clientes recurrentes de aproximadamente **38 % y 33 %**, respectivamente.
- **Junio** presenta el comportamiento más desfavorable, con una disminución de **4.4 % en clientes recurrentes** y de **9.4 % en la North Star**.
- **Conclusión:** existe una **asociación positiva** entre ambas variaciones, lo que respalda a los clientes recurrentes como un driver relevante de la North Star, aunque **no demuestra una relación causal**.

**Pregunta 7.** ¿La variación del driver acompaña la variación de la North Star? Indique un mes que se aparte de esa correspondencia y proponga una explicación verificable.

**Respuesta (máximo cuatro líneas):**

---

### **Correspondencia entre el driver y la North Star**

**Respuesta:** Sí, en general la variación de los **clientes recurrentes acompaña la variación de la North Star**, mostrando una asociación positiva entre ambas. Un caso que se aparta es **julio**, donde los clientes recurrentes aumentaron **1.0 %**, pero la North Star disminuyó **0.7 %**. Esto podría explicarse por una menor frecuencia de compra, que disminuyó **1.7 %**, situación que puede verificarse revisando ese indicador.

## Informe ejecutivo breve

1. **Objetivo estratégico:**  
2. **North Star y justificación:**  
3. **Hallazgo cuantitativo 1:**  
4. **Hallazgo cuantitativo 2:**  
5. **Hallazgo cuantitativo 3:**  
6. **Driver prioritario:**  
7. **Guardrails:**  
8. **Decisión empresarial recomendada:**  
9. **Limitación del dataset:**


## Ticket de salida

Elija una métrica de su proyecto integrador. Explique por qué es accionable y qué decisión cambiaría si disminuyera 20 %.

**Respuesta:**

### **Métrica accionable del proyecto integrador**

**Respuesta:** Elegiría la **tasa de retención de clientes**, porque permite evaluar si la empresa está logrando mantener a sus clientes y tomar acciones sobre fidelización y recuperación. Si disminuyera **20 %**, evaluaría inmediatamente estrategias de retención y campañas de reactivación para evitar una mayor pérdida de clientes.


## Referencias

- Chen, D. (2015). *Online Retail* [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5BW33  
- Croll, A., & Yoskovitz, B. (2013). *Lean Analytics*. O’Reilly Media.  
- Parmenter, D. (2020). *Key Performance Indicators* (4th ed.). Wiley.  
- Sharda, R., Delen, D., & Turban, E. (2024). *Business Intelligence, Analytics, Data Science, and AI* (5th ed.). Pearson.
